Linux: Fix vulkan init and GTK abstraction to run window-vulkan-demo

dougvj opened the pull request:
- Fix GTK abstractions on linux
- Fix Vulkan init support on linux
- Miscellaneous fixes for linux build/support
- Abstract the keyboard events so that nothing relies on hardcoded windows key codes. 

xenia-ui-window-vulkan-demo compiles and runs. 

xenia-app compiles, runs briefly and crashes with an assert as there is still missing threading functionality, however the GTK window is created correctly.

![xenia-ui-window-vulkan-demo](https://user-images.githubusercontent.com/6045454/34911560-3e8e49a0-f88a-11e7-8419-c8a71b025abf.png)

dougvj commented:
The tests were unhappy with some kind of DNS faliure on APT so I closed and opened it again to trigger a rerun

x1nixmzeng commented on src/xenia/ui/window_win.cc:
```diff
@@ -680,9 +680,200 @@ bool Win32Window::HandleMouse(UINT message, WPARAM wParam, LPARAM lParam) {
+    case VK_OEM_MINUS:
+      return KeyEvent::Key::kMinus;
+    case VK_OEM_PLUS:
+      return KeyEvent::Key::kEquals;
```
Is this correct?

dougvj replied:
Good question. That one confused me too. 

I think it is: 

https://www.gamedev.net/forums/topic/293214-virtual-keycode-hell---or-where-is-vk_equals/

x1nixmzeng commented:
Looks good to me! You even made some useful fixes for both platforms :+1:

x1nixmzeng commented:
Any comments on this PR @DrChat /@gibbed ?

DrChat commented:
@dougvj The changes look great at a glance! I'll review this in-depth as I get time.

DrChat commented on src/xenia/base/clock_posix.cc:
```diff
@@ -15,17 +15,17 @@
+  // Even if our resolution is less, CLOCK_MONOTONIC_RAW will always return
+  // ns values. Most of the time the resolution is pretty close to 1ns anyway
+  // due to the presence of hr timers, so just pass along ns ticks
+  return 1000000000ull;
```
Even though I can understand your reasoning - is there a specific reason you decided to hardcode this value rather than use `clock_getres`?

dougvj replied:
Good question. So the clock_getres always returns 1ns in Linux even if the clock doesn't actually support that resolution, so I think my thoughts at the time were why bother? But then again why did I bother to change it in the first place? I don't remember... I will revert this.

dougvj replied:
I noticed that the original frequency calculation was technically wrong even though it produced the right value, so I fixed that.

DrChat commented on src/xenia/ui/vulkan/vulkan_util.cc:
```diff
@@ -28,7 +28,7 @@ Version Version::Parse(uint32_t value) {
   version.major = VK_VERSION_MAJOR(value);
   version.minor = VK_VERSION_MINOR(value);
-  version.patch = VK_VERSION_PATCH(value);
+  version.patch = VK_VERSION_PATCH(VK_HEADER_VERSION);
```
This doesn't look quite right - this should still be parsing the input value.

dougvj replied:
I don't remember why I did this but didn't mean to commit it. Thanks for the catch.

jcash85 commented:
i have a issue with game not playing it will load up all the way through title screen then when it comes to game play it crashes dragon ball raging blast 1 and 2 and dragon ball ultimate tenkaichi worked two days ago

ananace commented on src/xenia/ui/imgui_drawer.cc:
```diff
@@ -117,23 +117,27 @@ void ImGuiDrawer::Initialize() {
+  io.KeyMap[ImGuiKey_Enter] = static_cast<int>(xe::ui::KeyEvent::Key::kEnter);
+  io.KeyMap[ImGuiKey_Escape] = static_cast<int>(xe::ui::KeyEvent::Key::kEsc);
+  io.KeyMap[ImGuiKey_A] = static_cast<int>(xe::ui::KeyEvent::Key::kA);
+  io.KeyMap[ImGuiKey_C] = static_cast<int>(xe::ui::KeyEvent::Key::kV);
```
I think this one is supposed to be `xe::ui::KeyEvent::Key::kC`

sl1pkn07 commented:
seems needs rebase

greetings

Margen67 commented:
@sl1pkn07 We're aware of this. Can you stop spamming PRs with this?

dougvj commented:
I will try to find time in the coming month to spend time on a rebase

Triang3l commented on src/xenia/ui/ui_event.h:
```diff
@@ -40,11 +40,107 @@ class FileDropEvent : public UIEvent {
-           bool modifier_alt_pressed, bool modifier_super_pressed)
+  // TODO(dougvj) The list of keycodes is probably incomplete but may be
+  // sufficient. Maybe there is a better way?
+  enum class Key {
```
It's preferable that we normalize keys to Windows virtual key codes, as those are sent to the game via XamInputGetKeystroke. This would also allow us to skip most of the translation for Windows host, maybe except for some fixups if needed to match the Xbox 360 controller keyboard better.

Triang3l commented:
`ui::KeyEvent::Key` superseded by `ui::VirtualKey` in https://github.com/xenia-project/xenia/commit/1cf12ec70b22c4eda28dea7fd8edf44900b098d6 — allows passing all Windows keys to XamInputGetKeystroke and UI subsystems without changes, but needs translation from GDK keys.

dougvj commented:
Rebased against master, borrowed commits from !1431. 

Builds are fixed and window runs, crashes when trying to run game.  Cpu tests fail. 

I dropped my GTK key abstraction I had in place, I will work on a mapping to the new `ui:VirtualKey`
